### ISSUE TITLE
(in-progress) frequent 5GHz disconnects

### ISSUE THREAD

**Issue by bjlockie** [morrownr/8821au#6]
I have an old laptop with a 2.4<ghz N card.
I want to use 5GHz and more speed is nice so I bought 
$ lsusb
Bus 002 Device 003: ID 2357:011e TP-Link AC600 wireless Realtek RTL8811AU [Archer T2U Nano]

I got it working BUT it keeps dropping association.
Does this look a problem with the USB2 ports?

I don't have control over which channel it uses (the router has no manual configuration of that, tplink Deco X60).
The 2.4GHz and 5GHZ ESSIDs are the same (the router...)

```
[  447.014923] RTW: Group handshake still in progress !!!
[  447.015331] RTW: 8812A: EDCA_BE_UL=0x5ea42b EDCA_BE_DL =0x5ea42b
[  447.088770] RTW: cfg80211_rtw_disconnect(wlx28ee520dce62) - Start to Disconnect
[  447.089480] RTW: rtw_sta_mstatus_disc_rpt(wlx28ee520dce62) - mac_id=0
[  447.090671] RTW: start_addr=1020, data_H:00010600, data_L:800000b5, macid=0, txrpt_h4b=10600
[  447.090674] RTW: MediaStatusRpt parm:00 00 00 
[  447.092671] RTW: wlx28ee520dce62 wakeup macid=0, ori reg_0x4d4=0x00000000
[  447.093918] RTW: rtw_hal_set_bssid wlx28ee520dce62- hw port -0 BSSID: 00:00:00:00:00:00
[  447.094668] RTW: rtw_set_rts_bw connect_to_8812=0,enable=1
[  447.095071] RTW: _rtw_disconnect...call rtw_indicate_disconnect
[  447.095101] RTW: rtw_free_assoc_resources-wlx28ee520dce62 tgt_network MacAddress=40:3f:8c:79:44:ab ssid=garlic
[  447.095114] RTW: ### Clean STA_(0) info ###
[  447.095120] RTW: rtw_release_macid(wlx28ee520dce62) if1, mac_addr:40:3f:8c:79:44:ab macid:0
[  447.095127] RTW: Free disconnecting network of scanned_queue
[  447.095137] RTW: rtw_cfg80211_unlink_bss(): cfg80211_unlink garlic!!
[  447.095163] RTW: rtw_cfg80211_indicate_disconnect(wlx28ee520dce62) ,reason = 0
[  447.095166] RTW: rtw_cfg80211_indicate_disconnect(wlx28ee520dce62) call cfg80211_disconnected, reason:0
[  447.095189] RTW: cfg80211_rtw_disconnect(wlx28ee520dce62) return 0
[  447.095212] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=0, addr=(null)
[  447.095215] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=1, addr=(null)
[  447.095217] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=2, addr=(null)
[  447.095219] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=3, addr=(null)
[  447.095221] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=4, addr=(null)
[  447.095224] RTW: cfg80211_rtw_del_key(wlx28ee520dce62) key_index=5, addr=(null)
[  447.095321] RTW: rtw_reg_notifier: NL80211_REGDOM_SET_BY_CORE
[  447.095381] RTW: rtw_reg_notifier: NL80211_REGDOM_SET_BY_USER alpha2:CA
[  447.095383] RTW: rtw_set_country(): not applied
[  447.096627] RTW: cfg80211_rtw_get_txpower
[  447.096857] RTW: rtw_reset_securitypriv(wlx28ee520dce62) - End to Disconnect
[  447.099219] RTW: rtw_ndev_notifier_call(wlx28ee520dce62) state:4
[  447.195814] RTW: cfg80211_rtw_scan(wlx28ee520dce62)
[  447.196310] RTW: SetHwReg8812A:(HW_VAR_CHECK_TXBUF)TXBUF Empty(1) in 0 ms
[  447.196557] RTW: wlx28ee520dce62 sleep m0=0x00000002, ori reg_0x4d4=0x00000000
[  447.199704] RTW: Switch ant band to ODM_BAND_2_4G
[  447.203335] RTW: HW_VAR_BASIC_RATE: 0x150->0x15f->0x15f
[  448.408329] RTW: Switch ant band to ODM_BAND_5G
[  448.411447] RTW: HW_VAR_BASIC_RATE: 0x150->0x150->0x150
[  449.031350] RTW: _rtw_phydm_pwr_tracking_rate_check i=0 if_tx_rate =0xc
[  449.031353] RTW: _rtw_phydm_pwr_tracking_rate_check tx_low_rate (unlinked to any AP)=0xc
[  451.039257] RTW: wlx28ee520dce62 wakeup m0=0x00000002, ori reg_0x4d4=0x00000002
[  451.039394] RTW: survey done event(43) band:0 for wlx28ee520dce62
[  451.039498] RTW: rtw_indicate_scan_done(wlx28ee520dce62)
[  451.043679] RTW: _rtw_phydm_pwr_tracking_rate_check i=0 if_tx_rate =0xc
[  451.043683] RTW: _rtw_phydm_pwr_tracking_rate_check tx_low_rate (unlinked to any AP)=0xc
[  451.062392] RTW: ==>rtw_ps_processor .fw_state(8)
[  451.062395] RTW: ==>ips_enter cnts:5
[  451.062396] RTW: nolinked power save enter
[  451.062397] RTW: ===> rtw_ips_pwr_down...................
[  451.062398] RTW: ====> rtw_ips_dev_unload...
[  451.073935] RTW: usb_read_port_cancel
[  451.077285] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.080536] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.083034] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.086659] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.089911] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.093161] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.096663] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.098894] RTW: usb_read_port_complete() RX Warning! bDriverStopped(False) OR bSurpriseRemoved(False)
[  451.099358] RTW: usb_write_port_cancel
[  451.099389] RTW: ==> rtl8812au_hal_deinit
[  451.100385] RTW:  hal_poweroff_8812au
[  451.114751] RTW: =====> _8051Reset8812(): 8051 reset success .
[  451.119014] RTW: SetHwReg: bMacPwrCtrlOn=0
[  451.119509] RTW: <=== rtw_ips_pwr_down..................... in 60ms
```
2.4GHz has been stable so far. :-)
It's 200mbps but I don't know if 150mbps @5GHz would be better.
Is 5GHz harder on USB?

**Comment by morrownr** [comment]
@bjlockie 

Hi, it is good to meet you. Let me take a look.

Quote: "The 2.4GHz and 5GHZ ESSIDs are the same (the router...)"

That is not good. You really need to find a way to give each band its own name. If you look in the README.md (front page) you will find a section on recommended router settings and having different names for the bands is on that list and for good reason. Find a way to make it happen.

Quote: "Is 5GHz harder on USB?"

Not that I have noticed. USB does work in mysterious ways sometimes so if you have more than one available USB port, try a different port. Sometimes adapters work better in certain points.

Do you use wavemon? If not, on Debian based systems...

$ sudo apt install wavemon

Then start it....

$ wavemon

It will give us some info about your link qualify and signal level as well as other information.

So, for now, your homework is to figure out how to get the settings in that wifi router lined out. The big issue is getting different names for the 2g and 5g bands.

Let me know how it goes.

P.S. Have you read this information?

https://github.com/morrownr/USB-WiFi

There is a section on the 8811au chipset. I just want you to be aware because the 8811au may be close to the end of its life. This driver and the one for the 8814au chipset are older drivers and I am having troubles keeping both going so the future of adapters with these chipsets is uncertain. I work on this and the other Realtek usb drivers to help my fellow Linux users but I really want Linux users to be aware that they may see better results with adapters that use alternatives to Realtek chipsets. I tell you this because if you just bought this adapter, maybe you have time to return it to get your money back.

**Comment by bjlockie** [comment]
Hi, thanks for doing this.

I bought this adapter for an old laptop with just USB2 because I wanted something really small with no external antenna.
It was supposed to just be temporary and I figured an older chipset would have a more stable/mature driver. :-)
I've had no luck making separate 2.4 and 5GHz ESSIDS.
I don't think it is possible with this mesh router from tplink.
There are complaints about other mesh routers from tplink on their forum back to 2019 and no fix yet so I don't think they will change it.
https://www.google.com/search?q=tplink+deco+x60+separate+2.4ghz+and+5ghz+sids&client=ubuntu&hs=6WR&channel=fs&ei=weR1YNG8JouEtQX9848I&oq=tplink+deco+x60+separate+2.4ghz+and+5ghz+sids&gs_lcp=Cgdnd3Mtd2l6EAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsANQ2E1Yo1FgwlxoAXABeACAAe0DiAHtA5IBAzQtMZgBAKABAaoBB2d3cy13aXrIAQjAAQE&sclient=gws-wiz&ved=0ahUKEwiRoa7O7vvvAhULQq0KHf35AwEQ4dUDCA0&uact=5
The recommended "solution" is to set up a 5GHz guest network with a different ESSID which is not a solution because the guest network doesn't have access to the local network (but it can access the internet).
The 8811AU driver works but only WITHOUT wep2 (no security).
WHY is separate ESSIDS recommended for different bands?

My Dad setup his old router only on 5GHz and connected to the new mesh router through a switch and I can connect to the old router on 5GHz with wep2 and it can see the internal network.

I also tried this:
I have an Alfa AWUS036ACM (Mediatek).
Anything with Mediatek is hard to find in Canada.
I had to buy it refurbished from Amazon.
I couldn't see any networks. 
```
[  193.313890] usb 2-4: New USB device found, idVendor=0e8d, idProduct=7612, bcdDevice= 1.00
[  193.313895] usb 2-4: New USB device strings: Mfr=2, Product=3, SerialNumber=4
[  193.313898] usb 2-4: Product: Wireless 
[  193.313902] usb 2-4: Manufacturer: MediaTek Inc.
[  193.313904] usb 2-4: SerialNumber: 000000000
[  193.512104] usb 2-4: reset high-speed USB device number 4 using ehci-pci
[  193.669906] mt76x2u 2-4:1.0: ASIC revision: 76120044
[  193.725151] mt76x2u 2-4:1.0: ROM patch build: 20141115060606a
[  193.889157] mt76x2u 2-4:1.0: Firmware Version: 0.0.00
[  193.889159] mt76x2u 2-4:1.0: Build: 1
[  193.889161] mt76x2u 2-4:1.0: Build Time: 201507311614____
[  194.740802] ieee80211 phy2: Selected rate control algorithm 'minstrel_ht'
[  194.768232] usbcore: registered new interface driver mt76x2u
[  194.798514] mt76x2u 2-4:1.0 wlx00c0caaa9c2b: renamed from wlan0
```
It worked fine as an access point on my raspberry pi.
It look pretty old too if the 2015 Build Time is accurate.

I wanted to buy a DLink DWA-X1850 - AX1800 Wi-Fi 6 USB 3.0 Adapter but it use a Realtek chipset according to DLink.
I'll wait to see if Realtek finally provides a Linux driver that can be included in the kernel.
Or maybe someone else will make a wifi6 USB device with good linux support.

**Comment by morrownr** [comment]
Good information but first...

What distro and version of Linux is on your laptop?

**Comment by bjlockie** [comment]
Ubuntu 20.10
lubuntu but almost the same.

**Comment by morrownr** [comment]
Okay, I am very familiar with Lubuntu and it is a very recent version.

The Alfa ACM is not really suited for use with a laptop but it is great with a RasPi. I use one with a RasPi4b as a wifi access point...

https://github.com/morrownr/7612u/blob/main/Bridged_Wireless_Access_Point-2.md

https://github.com/morrownr/7612u

I would not worry about the firmware date. We haven't had many changes for a few years. If something works, it works. I was just looking around my office. One of my systems has an 8 bit FM Radio ISA card in it that I use to listen to FM radio. Linux still has the driver in it. The card is from the mid-90s. My scanner is about 17 years old. It is plug and play with modern Linux distros. There is no driver for Windows 10. I don't think there was a driver for Windows 7. One of the really cool things about Linux is that it can allow us to continue to use perfectly good hardware for decades (if you get hardware with in-kernel drivers.)

Some adapters to consider...

https://www.ebay.com/p/27020163733

https://www.ebay.com/itm/363268018762

https://www.amazon.com/Socobeta-Network-Dual-Band-Wireless-Adapter/dp/B08MKNH2PJ

I maintain the following repo to help folks find adapters with in-kernel drivers:

https://github.com/morrownr/USB-WiFi

Concerning that new D-Link adapter. I will not recommend Linux users spend new money on Realtek adapters until such time as Realtek starts doing in-kernel drivers.

What you could do, since the that mesh router, doesn't want to cooperate, if take a RasPi 3b, 3b+ or 4b and that Alfa ACM and make yourself a wifi router/access point that will work with your laptop.

**Comment by morrownr** [comment]
Quote: "WHY is separate ESSIDS recommended for different bands?"

I had failed to answer this question previously so here goes. If both bands use the same ESSID, there will be switching between the bands and some drivers, I'm looking at you Realtek, do not handle this well.

I see the below thread concerning the router you have:

https://community.tp-link.com/us/home/forum/topic/161103

I consider this design to be bad. Personally, I won't buy tp-link or d-link products anymore. They do things that don't work well for Linux users and they don't properly support Linux users. This and they make heavy use of Linux in their products yet refuse to support Linux users properly.

I keep the following informational site for Linux users:

https://github.com/morrownr/USB-WiFi

Near the bottom of the front page, where I address the Realtek drivers, you will see that I specially recommend that Linux users do not buy adapters based on 2 chipsets: 8814au and 8811au. It is highly unlikely that either of these chipsets will ever be supported with in-kernel drivers and the existing drivers are both somewhat dated and the quality is questionable. I spend a lot of time in the source code of all of the drivers I have available here on this site and what I see in the source for these two drivers in particular is not good.

So, you are thinking "why does he keep these drivers available?" I have been using Linux since the mid-90's. Through the years, many Linux users have been good to me and I am just trying to pay that back. Many users have existing adapters when they come to Linux and that is what I consider the purpose of this site for. I, along with helpers, will try to keep these drivers going as best we can. On the other hand, if a user is asking what they should buy or if it is a user that just purchased an adapter and has time to take it back, I recommend they take it back and get something with in-kernel support. A good quality adapter with in-kernel support can provide many years of more or less trouble free service whereas fighting with Realtek drivers can take a toll on people.

**Comment by bjlockie** [comment]
My Dad bought a TPLInk mesh router as a spur of the moment purchase.

Are the 8811|8814/b|c drivers better written?
I chose the 8814au because it was a nano device.
I probably could have found a small Mediatek device that is small but it would have been wifi 4.
The more I read, the more I like Mediatek.
It looks like recent Atheros/Qualicom devices are not as open to Linux since Qualcomm bought Atheros.
What is the best wrtten wifi driver?

**Comment by morrownr** [comment]
Quote: "Are the 8811|8814/b|c drivers better written?"

My opinion is that of the Realtek drivers that I have available here, there is only one that I would call a good driver and that is the one for the rtl8812au. It is by far the best driver for usb adapters from Realtek. I'm not aware of any adapters that are small that use that chipset though.

Quote: "I probably could have found a small Mediatek device that is small but it would have been wifi 4."

Just because you are not aware of small Mediatek based adapters does not mean they are not available <grin>

https://www.ebay.com/p/27020163733

Go to...

https://github.com/morrownr/USB-WiFi

...and go to the AC600 section. The last 6 adapters listed under Mediatek mt7610u chipset are small adapters. They are all WiFi 5.

Quote: "What is the best wrtten wifi driver?"

I need to keep my answer limited to USB WiFi as that is where my experience is. My opinion is that the best USB WiFi drivers are the ones for the mt7612u chipset (AC1200) and the mt7610u chipset (AC600). Here is a quote from a person that filed an issue over in the USB-WIFI repo...

"I'm starting to enjoy the MediaTek chipset, I haven't found any bugs with the driver yet."

He was talking about an Alfa adapter based on the my7612u. He had originally posted an Issue because he was having trouble with hostapd. He needed the proper settings for ht= and vht=. I pointed him to the correct settings and off he went.

I have adapters based on both the mt7612u chipset and the mt7610u chipset. It is hard to come up with bugs but if I have to dig really hard to come up with something...

- The source from the mt7612u chipset is written to look for its firmware in /lib/firmware whereas the rest of the Mediatek drivers are written to look for their firmware in /lib/firmware/mediatek. Not a big deal and it does not cause problems but it violates my sense of consistency. I'm going to point it out to the Mediatek devs when I get a chance.

- The RasPiOS does not include the firmware for the mt7610u. The firmware for the mt7612u is there but not the mt7610u. I reported this recently to the RasPiOS devs. Maybe they can fix that. It is easy for users to fix but it is the job of distros to include firmware.

- Neither driver currently supports onboard LEDs. If the support was there, I would be looking for a way to turn it off so the devs may have done this intentionally. I do see code that seems to indicate that they may be looking at adding this support. If I get time, I may investigate and see if I can add the support myself. I added the LED configuration support to these Realtek drivers so maybe I can do it.

As you can see, I'm pretty picky and I just have not run onto any problems of any consequence with the current Mediatek drivers and these drivers are in the kernel and they are standards compliant.

**Comment by morrownr** [comment]
Oh, something I just thought of. Given a small adapter using the 5ghz band, you do realize you need to be pretty close to your wifi router/access point? Don't expect it to give you fast throughput through 2 walls and over 40 feet. Not going to happen. 5ghz, being a higher frequency, can't penetrate walls as good as 2ghz can. That is physics getting involved. The very small antennas don't help much either. I just say this make sure what you are asking for is what is going to meet your needs.

**Comment by morrownr** [comment]
Quote: "The RasPiOS does not include the firmware for the mt7610u. The firmware for the mt7612u is there but not the mt7610u. I reported this recently to the RasPiOS devs. Maybe they can fix that. It is easy for users to fix but it is the job of distros to include firmware."

I got a reply from a RasPiOS dev that pointed me to a Debian bug. Said bug has already been corrected, When RasPiOS rebases on the next Debian stable, this issue goes away. Arg! Most distros that are based on Debian, such as Ubuntu, correct out issues like this on their own without waiting for Debian because waiting on Debian is like watching glaciers flow. That bug was fixed in the package that contains it around 1 year ago and we have to wait for the next stable release to actually see the fix!?!

**Comment by bjlockie** [comment]
It is weird but I get 5GHz from the basement way better than 2.4 from the basement was.
This laptop only has USB2 and it says I get 434000 kbps with this Realtek.

**Comment by morrownr** [comment]
It is not clear to me what you are saying. Could you rephrase?

**Comment by bjlockie** [comment]
The router in the basement connects to the main floor better on 5GHz better than 2.4GHz ever did.

**Comment by morrownr** [comment]
Probably because of having a better antenna,

Where are we on this issue?
